common/test/TestLogging.cs

```diff
@@ -2,39 +2,115 @@
+        public const string Language = "C#";
+        public const string Service = "IotHub";
+        private readonly object initLock = new object();
+
+        // Client to log to application insights
+        private static TelemetryClient _telemetryClient;
+
+        // Unique ID for all tests of a run.
+        private static string _testRunId;
+
+        public TestContext Context { get; set; }
+        public string TargetFramework { get; set; }
 
         private TestLogging()
         {
+            // Thread-safe initialization of the telemetry client and run Id.
+            lock (initLock)
+            {
+                if (_telemetryClient == null)
+                {
+                    string iKey = Environment.GetEnvironmentVariable("E2E_IKEY");
+                    if (!string.IsNullOrWhiteSpace(iKey))
+                    {
+                        var config = new TelemetryConfiguration
+                        {
+                            InstrumentationKey = iKey,
+                        };
+                        _telemetryClient = new TelemetryClient(config);
+                        _testRunId = Guid.NewGuid().ToString();
+                    }
+                }
+            }
         }
 
         private const string NullInstance = "(null)";
-        
+
         public static string IdOf(object value) => value != null ? value.GetType().Name + "#" + GetHashCode(value) : NullInstance;
 
         public static int GetHashCode(object value) => value?.GetHashCode() ?? 0;
 
-        public static TestLogging GetInstance()
+        public static TestLogging GetInstance(TestContext testContext = null)
         {
-            return s_instance;
+            var logger = new TestLogging();
+
+            if (_telemetryClient != null)
+            {
+                // Set logger properties for the current test.
+                var targetFramework = (TargetFrameworkAttribute)Assembly.GetExecutingAssembly()
+                .GetCustomAttributes(typeof(TargetFrameworkAttribute), false)
+                .SingleOrDefault();
+                logger.Context = testContext;
+                logger.TargetFramework = targetFramework.FrameworkName;
+            };
+
+            return logger;
         }
 
-        public void WriteLine(string message)
+        public void WriteLine(string message, SeverityLevel severity = SeverityLevel.Information, IDictionary<string, string> propertyBag = null, [CallerMemberName] string caller = null)
         {
+            // Log to event source
             EventSourceTestLogging.Log.TestMessage(message);
+
+            // Log to Application insights
+            if (_telemetryClient != null)
+            {
+                if (propertyBag == null)
+                {
+                    propertyBag = new Dictionary<string, string>();
+                }
+
+                // Add common properties
+                propertyBag.Add(LoggingPropertyNames.TestRunId, _testRunId);
```

Comment:
Why not have a persistent property bag you merge with, like we did in DTL?

Reply:
I had it that way originally but there were hardly any constant properties so decided to create one on the fly instead of having another dictionary. Do you think having a persistent property bag is better? In this case, every test has its own property bag other than 2-3 dimensions.

Reply:
If you recall for DTL, we had a property bag for the initial logger.

Then each operation (for here that would be a test class or a test) would clone that, and add its own persistent properties.

Then each call to Log or Event could add additional properties, but that is rare.

So in the log method, you'd check if there are any custom properties, if so you'd merge persistent and transient into a new dictionary, and log that. If not, you'd just pass the persistent properties. That means, in most log cases, you'd not create a dictionary each time.

Reply:
True but other than the RunId, everything else is per test. To keep it simple and not have to clone, I just created it this way. We can surely do that if we think this is complicated or if it makes the code more generic.